### PR TITLE
feat: split KV cache for MiMo-V2-Flash SWA layers

### PR DESCRIPTION
**Pull request by MokusMokun**
## Summary

- Add split-KV ragged paged attention kernel (v3) that stores K and V caches separately with different head dimensions (K=256, V=128 for MiMo-V2-Flash)
- Fix `can_use_split_kv_cache` to check **per-device** head count instead of total heads, ensuring full-attention layers (per-device kv_heads=1) fall back to fused KV while SWA layers (per-device kv_heads=2) use split KV
- Move V padding from model layer to FlashAttention backend, enabling per-layer split/fused dispatch

## Key changes

| File | Change |
|---|---|
| `ragged_paged_attention_v3_split.py` | New Pallas kernel with 5D packed VMEM layout, strided load optimization, with Public API wrapper + reference implementation |
| `memory_pool.py` | `can_use_split_kv_cache` checks per-device heads via mesh; `MHATokenToKVPool` supports split buffers; `SWAKVPool` handles mixed split/fused pools |
| `model_runner.py` | Pass `v_head_dim` and `mesh` through init chain; handle mixed tuple/array in `_set_kv_cache_after_forward` |
| `flashattention_backend.py` | Per-layer `_call_split_kv` / `_call_fused_kv` dispatch; V padding moved here for fused path |
| `mimo_v2_flash.py` | Remove model-level V padding (now handled by backend) |

## KV Cache memory savings

Measured on GKE v6e-16 (4 nodes × 4 chips). KV cache layout is determined by model architecture, independent of code optimizations.

| Pool | Baseline (all fused) | Split-KV (this PR) | Saved |
|---|---:|---:|---|
| SWA 39 layers | 70.99 GB | 53.25 GB (K=35.50 + V=17.75) | **-17.75 GB (-25%)** |
| Full Attn 9 layers | 40.82 GB | 40.82 GB (fused fallback) | 0 |
| **Total KV cache** | **111.81 GB** | **94.07 GB** | **-17.74 GB (-15.9%)** |

Server log confirmation:
```
Creating split K/V buffers for 39 layers (head_num=8, k_dim=256, v_dim=128)
Disabling split KV cache for local head layout: head_num=4, k_dim=256, v_dim=128
Creating fused KV buffers for 9 layers
```

## E2E performance

Split-KV branch (`31dc0ca1`) is based on official baseline commit (`cef4a181`). Performance compared against official baseline data from `mimo_v2_flash_epic_baseline_final.md`.

Both runs: `--mem-fraction-static 0.95 --max-running-requests 128 --max-concurrency 64`

| Metric | Baseline (`cef4a181`) | Split-KV (`31dc0ca1`) | Delta |
|---|---:|---:|---|
| Output throughput (tok/s) | 654.86 | 651.32 | **-0.5%** |
| Peak output throughput (tok/s) | 2560 | 2624 | **+2.5%** |
| Median ITL (ms) | 29.40 | 29.18 | **-0.7%** |
| P99 ITL (ms) | 32.82 | 38.56 | +17.5% |
| Mean TTFT (ms) | 34062 | 34466 | +1.2% |

## E2E accuracy (GSM8K, 5 runs)

| Run | Split-KV | Baseline |
|---|---:|---:|
| Run 1 | 0.9166 | 0.9105 |
| Run 2 | ~0.9155 (1318/1319) | 0.9219 |
| Run 3 | 0.9158 | 0.9356 |
| Run 4 | 0.9204 | - |
| Run 5 | 0.9333 | - |
| **Mean** | **0.9203** (5-run) | **0.9227** (3-run) |

Delta: **-0.24%**, well within sampling variance (baseline range: 91.05–93.56, split-KV range: 91.55–93.33).

## Test plan

- [x] Single-chip kernel correctness: 11/11 tests pass (decode/prefill/SWA/attention_sink)
- [x] shard_map tp=4 SWA config (per-device=2): PASS
- [x] E2E performance benchmark vs official baseline: throughput/ITL within 1%
- [x] E2E accuracy (GSM8K): 5-run mean 92.03 vs baseline 92.27 (-0.24%), no regression
- [x] Server starts with baseline params (mem=0.95, max-running=128)
- [x] Server log confirms: SWA 39 layers → split, Full Attn 9 layers → fused
- [x] KV cache memory: -17.75 GB (-25% SWA, -15.9% total)

## 下一步
1. 5D packed split kernel（继承自 [PR#217](https://github.com/primatrix/sglang-jax/pull/217) 的 bf16 tiling 优化）要求 per-device kv_heads >= 2。Full attention 层在 tp=4 下 per-device kv_heads=1，无法使用 split KV，只能回退到 fused（V 被 pad 到 256 与 K 对齐）。这 9 层的 V cache 未能从 256→128 缩减，是目前 split KV 内存收益的上限瓶颈
2. 修复 profile_max_num_token 正确估算混合 split/fused cell_size（89,088 → 69,120 bytes/token，-22.4%），使 max_total_num_tokens 增加 ~29%

### CODE REVIEW

File: benchmark/kernels/flash_attention/utils.py [hunk 1]
```diff
@@ -172,3 +167,45 @@ def create_decode_uniform_data(
...
+
+
+# ---------------------------------------------------------------------------
+# Split KV helpers (separate K and V with different head_dim)
+# ---------------------------------------------------------------------------
+
+from sgl_jax.srt.kernels.ragged_paged_attention.util import get_dtype_packing, align_to
+
+
+def create_split_qkv_data(
```
Comment (pathfinder-pf) [inline benchmark/kernels/flash_attention/utils.py]: What is the purpose of this function?

File: python/sgl_jax/srt/model_executor/model_runner.py [hunk 2]
```diff
@@ -701,10 +717,21 @@ def _set_kv_cache_after_forward(self, layers_kv_fused):
...
-                jax.device_put(layer_kv_fused, target_sharding)
-                for layer_kv_fused in layers_kv_fused
-            ]
+            # Handle mixed split/fused layers: each element is either a tuple
+            # (k_cache, v_cache) from split-KV layers or a single array from
+            # fused-KV layers.
+            resharded = []
+            for item in layers_kv_fused:
+                if isinstance(item, tuple):
+                    resharded.append(
```
Comment (pathfinder-pf) [inline python/sgl_jax/srt/model_executor/model_runner.py:729]: please assert tuple size is 2

File: python/sgl_jax/srt/model_executor/model_runner.py [hunk 3]
```diff
@@ -382,31 +383,43 @@ def profile_max_num_token(self, total_device_memory: int):
             raise RuntimeError("Not enough memory. Please try to increase --mem-fraction-static.")
 
         # head_dim/v_head_dim handling
-        # V is padded to head_dim in model layer for fused KV cache
+        # With split KV, V uses v_head_dim; with fused KV, V is padded to head_dim
         head_dim = self.model_config.head_dim
+        v_head_dim = self.model_config.v_head_dim
```
Comment (pathfinder-pf) [inline python/sgl_jax/srt/model_executor/model_runner.py]: What should be done if the model does not have v_head_dim?

File: python/sgl_jax/srt/mem_cache/memory_pool.py [hunk 4]
```diff
@@ -49,6 +49,29 @@ def merge_kv(k: jax.Array, v: jax.Array) -> jax.Array:
...
+    the 'tensor' axis size; otherwise it falls back to ``head_num`` directly.
+    """
+    if v_head_dim == head_dim:
+        return False
+    per_device_heads = head_num
+    if mesh is not None:
+        tp_devices = mesh.shape.get("tensor", 1)
+        if tp_devices > 0:
+            per_device_heads = head_num // tp_devices
+    return per_device_heads >= 2
```
Comment (pathfinder-pf) [inline python/sgl_jax/srt/mem_cache/memory_pool.py]: per_device_heads % 2 == 0 maybe better

File: python/sgl_jax/srt/mem_cache/memory_pool.py [hunk 5]
```diff
@@ -249,10 +272,22 @@ def __init__(
...
+        v_head_dim: int = -1,
     ):
         super().__init__(size, page_size, dtype, layer_num, mesh, start_layer, end_layer)
         self.head_num = head_num
         self.head_dim = head_dim
+        self.v_head_dim = v_head_dim if v_head_dim > 0 else head_dim
+        self.split_kv = can_use_split_kv_cache(
+            self.head_num, self.head_dim, self.v_head_dim, mesh=self.mesh
+        )
+        if self.v_head_dim != self.head_dim and not self.split_kv:
```
Comment (pathfinder-pf) [inline python/sgl_jax/srt/mem_cache/memory_pool.py:284]: Does this mean we're still using the fuse for the case tp=4 and head = 4?

**Comment by pathfinder-pf** [comment]
please add e2e accuracy test for MiMo-V2-Flash and other model.

File: benchmark/kernels/flash_attention/utils.py (hunk 1 above)
Comment (MokusMokun) [reply benchmark/kernels/flash_attention/utils.py]: helper function for previous removed tests, now is redundant, has been deleted

File: python/sgl_jax/srt/mem_cache/memory_pool.py (hunk 5 above)
Comment (MokusMokun) [reply python/sgl_jax/srt/mem_cache/memory_pool.py:284]: yes (full attention still use fuse_256), you can check the PR description for this

File: python/sgl_jax/srt/model_executor/model_runner.py (hunk 2 above)
Comment (MokusMokun) [reply python/sgl_jax/srt/model_executor/model_runner.py:729]: done

File: python/sgl_jax/srt/model_executor/model_runner.py (hunk 3 above)
Comment (MokusMokun) [reply python/sgl_jax/srt/model_executor/model_runner.py]: resolved using getattr()

File: python/sgl_jax/srt/mem_cache/memory_pool.py (hunk 4 above)
Comment (MokusMokun) [reply python/sgl_jax/srt/mem_cache/memory_pool.py]: addtional check added

**Comment by MokusMokun** [comment]
> please add e2e accuracy test for MiMo-V2-Flash and other model.

the e2e test has been done for MiMo-V2-Flash as above
here is the test result for Qwen3-8B （instruct模型）
使用模型：https://huggingface.co/Qwen/Qwen3-8B

## Accuracy

### MMLU (run_eval.py, 64 examples, 3 runs)

| Run | Score | Social Sciences | Other | Humanities | STEM |
|-----|-------|-----------------|-------|------------|------|
| 1   | 0.719 | 0.857           | 0.750 | 0.652      | 0.636|
| 2   | 0.766 | 0.786           | 0.813 | 0.783      | 0.636|
| 3   | 0.766 | 0.857           | 0.750 | 0.783      | 0.636|
| **Avg** | **0.750** | 0.833     | 0.771 | 0.739      | 0.636|


**0-shot (3 runs)**

| Run | Score |
|-----|-------|
| 1   | 0.859 |
| 2   | 0.853 |
| 3   | 0.865 |
| **Avg** | **0.859** |

波动范围: 0.853 ~ 0.865 (delta 1.4%)

**4-shot CoT (3 runs)**

| Run | Score |
|-----|-------|
| 1   | 0.920 |
| 2   | 0.923 |
| 3   | 0.924 |
| **Avg** | **0.922** |

波动范围: 0.920 ~ 0.924 (delta 0.4%)

| 模式 | 模型 | Score (avg) | 与官方对比 |
|------|------|-------------|-----------|
| 0-shot | Instruct | 0.859 | -3.9% |
| 4-shot CoT | Instruct | **0.922** | **+2.4%** |
| 4-shot CoT | **Base** | 0.563 | -33.5% |
| 官方 (4-shot CoT) | Base | 0.898 | baseline |

> **Instruct model 4-shot CoT (0.922)** 超过官方 base model 数字（对于qwen-8b-base的精度复现过程见下）

## Benchmark

### ISL=1024, OSL=1024

| Batch | Median TTFT (ms) | Median ITL (ms) | Input Throughput (tok/s) | Output Throughput (tok/s) | Peak Output (tok/s) |
|-------|------------------|-----------------|--------------------------|---------------------------|---------------------|
| 8     | 93.41            | 6.87            | 1147                     | 1147                      | 1232                |
| 16    | 155.39           | 7.97            | 1953                     | 1953                      | 2032                |
| 32    | 285.59           | 9.74            | 3139                     | 3139                      | 3456                |
| 64    | 527.52           | 14.02           | 4342                     | 4342                      | 4864                |
| 128   | 1003.88          | 18.51           | 6297                     | 6297                      | 7808                |
| 256   | 2060.63          | 26.39           | 8400                     | 8400                      | 10537               |

### ISL=4096, OSL=1024

| Batch | Median TTFT (ms) | Median ITL (ms) | Input Throughput (tok/s) | Output Throughput (tok/s) | Peak Output (tok/s) |
|-------|------------------|-----------------|--------------------------|---------------------------|---------------------|
| 8     | 350.71           | 6.96            | 4248                     | 1062                      | 1192                |
| 16    | 644.89           | 8.12            | 6883                     | 1721                      | 2032                |
| 32    | 1229.49          | 10.66           | 9932                     | 2483                      | 3099                |
| 64    | 2415.64          | 16.65           | 12037                    | 3009                      | 3968                |
| 128   | 4745.21          | 27.94           | 12616                    | 3154                      | 4440                |
| 256   | 41905.86         | 28.00           | 12949                    | 3237                      | 4440                |

### ISL=8192, OSL=1024

| Batch | Median TTFT (ms) | Median ITL (ms) | Input Throughput (tok/s) | Output Throughput (tok/s) | Peak Output (tok/s) |
|-------|------------------|-----------------|--------------------------|---------------------------|---------------------|
| 8     | 715.74           | 7.29            | 7392                     | 924                       | 1144                |
| 16    | 1418.26          | 8.46            | 11362                    | 1420                      | 1936                |
| 32    | 2845.04          | 13.46           | 13473                    | 1684                      | 2464                |
| 64    | 5701.66          | 23.17           | 14899                    | 1862                      | 2880                |
| 128   | 40449.11         | 24.07           | 14759                    | 1845                      | 2838                |
| 256   | 113746.58        | 24.36           | 14346                    | 1793                      | 2856                |

**Comment by MokusMokun** [comment]
### Qwen3-8B-Base model 精度情况

使用模型：https://huggingface.co/Qwen/Qwen3-8B-Base
Base model 3 轮 chat API 4-shot CoT 仅得 0.563（官方 0.898）。逐步排查：

**尝试 1: 改用 Completion API + stop tokens + `####` answer extraction**

simple-evals `--use-completion` 模式，4-shot CoT，3 轮全量：

| Run | Score |
|-----|-------|
| 1   | 0.735 |
| 2   | 0.754 |
| 3   | 0.745 |
| **Avg** | **0.745** |

从 0.563 提升到 0.745（+18.2%），但仍低于官方 0.898。

**尝试 2: lm-eval-harness（标准 8-shot CoT）**

使用 `lm_eval --model local-completions --tasks gsm8k_cot`，标准 8-shot CoT prompt：

| Metric | Score |
|--------|-------|
| flexible-extract | **0.7324** |
| strict-match | 0.6861 |

仍低于官方 ~16%。

**结论**: 差距可能来自 Qwen 官方使用自家评估框架（非 lm-eval-harness），或 generation 参数（temperature/top_p）不同。对齐官方 base model 数字需要复现其完整评估 pipeline，超出当前 scope。
